contracts/most/Utmost.sol

```diff
@@ -0,0 +1,157 @@
+ *      auxiliary chain.
+ */
+contract Utmost is MasterCopyNonUpgradable, GenesisUtmost, UtilityToken {
+
+    /* Events */
+
+    /** Emitted whenever the ERC20 Utmost is converted to base coin Utmost. */
+    event TokenUnwrapped(
+        address indexed _account,
+        uint256 _amount
+    );
+
+    /** Emitted whenever the base coin Utmost is converted to ERC20 Utmost. */
+    event TokenWrapped(
+        address indexed _account,
+        uint256 _amount
+    );
+
+
+    /** Constants */
+
+    /** Token Symbol */
+    string public constant TOKEN_SYMBOL = "UM";
+
+    /** Token Name */
+    string public constant TOKEN_NAME = "Utmost";
+
+    /** Token Decimal */
+    uint8 public constant TOKEN_DECIMALS = 18;
+
+
+    /* External Functions */
+
+    /**
+     * @notice Sets up the genesisTokenSupply and calls UtilityToken.setup.
+     *
+     * @dev This function must be called only once. This check is done in setup
+     *      function of UtilityToken.
+     *
+     * @param _genesisTokenSupply Initial token supply.
+     * @param _consensusCogateway ConsensusCogateway contract address.
+     *
+     */
+    function setup(
+        uint256 _genesisTokenSupply,
+        address _consensusCogateway
+    )
+        external
+    {
+        UtilityToken.setup(
+            TOKEN_SYMBOL,
+            TOKEN_NAME,
+            TOKEN_DECIMALS,
+            _genesisTokenSupply,
+            _consensusCogateway
+        );
+
+        genesisTotalSupply = _genesisTokenSupply;
+    }
+
+    /**
+     * @notice Unwrap converts ERC20 Utmost to base coin.
+     *
+     * @dev  This contract's base coin balance must always be greater than
+     *       unwrap amount.
+     *
+     * @dev Function requires:
+     *      - Amount should be non-zero
+     *      - Caller Utmost token balance should be greater than amount
+     *
+     * @param _amount Amount of ERC20 Utmost token to convert to base coin.
+     */
+    function unwrap(
+        uint256 _amount
+    )
+        external
+        returns (bool success_)
+    {
+        require(
+            _amount > 0,
+            "Amount is zero."
+        );
+
+        require(
+            _amount <= balances[msg.sender],
+            "Insufficient balance."
+        );
+
+        assert(address(this).balance >= _amount);
+
+        transferBalance(msg.sender, address(this), _amount);
+
+        msg.sender.transfer(_amount);
+
+        emit TokenUnwrapped(msg.sender, _amount);
+
+        success_ = true;
+    }
+
+    /**
+     * @notice Wrap converts base coin to ERC20 Utmost.
+     *
+     * @dev The ERC20 OST balance of contract should always be greater than the
+     *      received payable amount.
+     *
+     * @dev Function requires:
+     *      - msg.value should be non-zero
+     *
+     * @return success_ `true` if wrap was successfully progressed.
+     */
+    function wrap()
+        external
+        payable
+        returns (bool success_)
```

Comment:
Similarly, do we need to return anything?
Also, let us think if mutex should be added.

Reply:
Same comment as above